File: modules/services/versioned_admin_view_models.py
"""Deterministic UI state helpers for versioned administration pages."""

from __future__ import annotations

from typing import Any, Iterable, Mapping


def lifecycle_actions(
    status: str, *, has_published_version: bool
) -> dict[str, bool]:
    """Return lifecycle action availability for a template."""
    return {
        "can_activate": status == "inactive" and has_published_version,
        "can_deactivate": status == "active",
        "can_archive": status == "inactive",
        "can_edit": status != "archived",
        "can_publish": status != "archived",
    }


def version_label(version: Mapping[str, Any], *, kind: str) -> str:
    """Return an exact, human-readable pipeline/schema version label."""
    if kind not in {"pipeline", "schema"}:
        raise ValueError("Version label kind must be pipeline or schema.")
    if kind == "schema":
        prefix = version.get("schema_key") or version.get("template_name")
    else:
        prefix = version.get("template_key") or version.get("name")
    number = version.get("version_number") or "—"
    hash_suffix = (
        f" · {str(version['content_hash'])[:10]}"
        if version.get("content_hash")
        else ""
    )
    return f"{prefix or 'Template'} · v{number}{hash_suffix}"


def schema_selector_options(
    versions: Iterable[Mapping[str, Any]],
) -> list[dict[str, Any]]:
    """Return stable exact-version options without choosing one implicitly."""
    return [
        {
            "value": str(version["id"]),
            "label": version_label(version, kind="schema"),
            "content_hash": version.get("content_hash"),
            "version_number": version.get("version_number"),
        }
        for version in versions
    ]


def group_findings(
    findings: Iterable[Mapping[str, Any]],
) -> dict[str, list[dict[str, Any]]]:
    """Group validation findings by severity without changing order."""
    grouped: dict[str, list[dict[str, Any]]] = {}
    for finding in findings:
        item = dict(finding)
        grouped.setdefault(str(item.get("severity") or "error"), []).append(item)
    return grouped


def revision_conflict_view(detail: Mapping[str, Any]) -> dict[str, Any]:
    """Return explicit stale-writer state without selecting a replacement."""
    return {
        "message": str(detail.get("message") or "The draft changed on the server."),
        "current": detail.get("current"),
        "reload_required": detail.get("current") is not None,
        "overwrite_allowed": False,
    }

File: web/static/js/versioned_admin_view_models.js
(function () {
    "use strict";

    function lifecycleActions(status, hasPublishedVersion) {
        return {
            canActivate: status === "inactive" && Boolean(hasPublishedVersion),
            canDeactivate: status === "active",
            canArchive: status === "inactive",
            canEdit: status !== "archived",
            canPublish: status !== "archived",
        };
    }

    function versionLabel(item, kind) {
        const prefix = kind === "schema" ? (item.schema_key || item.template_name) : (item.template_key || item.name);
        const hash = item.content_hash ? ` · ${String(item.content_hash).slice(0, 10)}` : "";
        return `${prefix || "Template"} · v${item.version_number || "—"}${hash}`;
    }

    function groupFindings(findings) {
        return (findings || []).reduce((groups, finding) => {
            const severity = finding.severity || "error";
            (groups[severity] = groups[severity] || []).push(finding);
            return groups;
        }, {});
    }

    function conflictMessage(error) {
        const detail = error && error.detail;
        if (detail && typeof detail === "object" && detail.current) {
            return {
                message: detail.message || "The draft changed on the server.",
                current: detail.current,
                reloadRequired: true,
            };
        }
        return { message: (detail && detail.message) || detail || error.message || "Request failed.", reloadRequired: false };
    }

    window.DocFlowVersionedAdmin = {
        lifecycleActions,
        versionLabel,
        groupFindings,
        conflictMessage,
    };
})();
